content/post/word_ladder.md

```diff
@@ -0,0 +1,689 @@
+		fmt.Println(l)
+	}
+```
+
+[This](/code/word_ladders/words-1a.go) now gives us all the shortest word ladders from "head" to "tail".
+```
+$ xtime ./words-1a -first head -last tail </usr/share/dict/words
+[head read reid raid rail tail]
+[head heal neal neil nail tail]
+[head heal teal tell tall tail]
+[head held hell tell tall tail]
+[head heal hell tell tall tail]
+[head held hell hall tall tail]
+[head heal hell hall tall tail]
+[head hear heir hair hail tail]
+[head held hell hall hail tail]
+[head heal hell hall hail tail]
+0.07u 0.00s 0.07r 8876kB ./words-1a -first head -last tail
+```
+
+Getting back to the single path problem, the next step is to see what we can do to improve the performance of the program. The observation that can drive this is that when we are constructing the graph in the naive implementation and the `wordGraph` wrapping of that approach, we perform neighbourhood calculations even for words that are unreachable from our doublet. This work will never be used. Depending on the word length and the doublet we choose this can make a reasonable difference.
+
+The main function remains the same, but `wordGraph` is replaced with the following (and goimports adjusts out import list) so that we determine neighbourhoods lazily as we reach them during the shortest path search.
+
+```
+type wordGraph struct {
+	n     int
+	words []string
+	ids   map[string]int64
+}
+
+func newWordGraph(n int) wordGraph {
+	return wordGraph{n: n, ids: make(map[string]int64)}
+}
+
+func (g *wordGraph) include(word string) {
+	if len(word) != g.n || !isWord(word) {
+		return
+	}
+	word = strings.ToLower(word)
+	if _, exists := g.ids[word]; exists {
+		return
+	}
+	g.ids[word] = int64(len(g.words))
+	g.words = append(g.words, word)
+}
+
+func isWord(s string) bool {
+	for _, c := range []byte(s) {
+		if lc(c) < 'a' || 'z' < lc(c) {
+			return false
+		}
+	}
+	return true
+}
+
+func lc(b byte) byte {
+	return b | 0x20
+}
+
+func (g wordGraph) nodeFor(word string) graph.Node {
+	id, ok := g.ids[word]
+	if !ok {
+		return nil
+	}
+	return node{word, id}
+}
+
+func (g wordGraph) From(id int64) graph.Nodes {
+	if uint64(id) >= uint64(len(g.words)) {
+		return graph.Empty
+	}
+	return newNeighbours(g.words[id], g.ids)
+}
+
+func (g wordGraph) Edge(uid, vid int64) graph.Edge {
+	if uid == vid {
+		return nil
+	}
+	if uint64(uid) >= uint64(len(g.words)) {
+		return nil
+	}
+	if uint64(vid) >= uint64(len(g.words)) {
+		return nil
+	}
+	u := g.words[uid]
+	v := g.words[vid]
+	d := hamming(u, v)
+	if d != 1 {
+		return nil
+	}
+	return edge{f: node{u, uid}, t: node{v, vid}}
+}
+
+func hamming(a, b string) int {
+	if len(a) != len(b) {
+		panic("word length mismatch")
+	}
+	var d int
+	for i, c := range []byte(a) {
+		if c != b[i] {
+			d++
+		}
+	}
+	return d
+}
+
+type neighbours struct {
+	word string
+	ids  map[string]int64
+	j    int
+	d    byte
+	buf  []byte
+	curr graph.Node
+}
+
+// completely deterministic - so only gives on solution
```

Comment:
s/on/one but probably just remove the comment.

Reply:
The entirety of the comment or just the portion after the dash?

Reply:
Replaced with a proper comment.